File: tests/refresh_test.ts
/**
 * Tests for LyXServer refresh helpers.
 *
 * These tests verify the safety contracts of refreshPreStep
 * WITHOUT requiring a running LyX instance. The socket client (sendLyxCommands)
 * returns false when no LyX is running, which is the path we test here.
 */

import { assert, assertEquals } from "@std/assert";
import { refreshPreStep } from "../src/cli.ts";
import { buildPipeCommand, filterResponses } from "../src/lyxserver.ts";

Deno.test("Refresh - save-reload pre-step returns a status", { timeout: 10000 }, async () => {
  // The pre-step returns one of "ok" | "disconnect" | "unconfirmed" | "error"
  // (test_report_38 F1 Option A): a genuine disconnect aborts, an unconfirmed
  // (dispatched but lost) save proceeds with a warning, a confirmed error
  // aborts. This test verifies the function doesn't crash and returns a valid
  // status; the actual value depends on whether LyX happens to be running.
  const status = await refreshPreStep("/tmp/test.lyx", "save-reload");
  assert(
    ["ok", "disconnect", "unconfirmed", "error"].includes(status),
    `refreshPreStep must return a status string, got: ${status}`,
  );
});

Deno.test("Refresh - reload mode has no pre-step", { timeout: 5000 }, async () => {
  // reload mode intentionally discards unsaved edits — no pre-step needed.
  // If the guard were removed, this would call sendLyxCommands, which returns
  // false without LyX, and the test would fail.
  const ok = await refreshPreStep("/tmp/test.lyx", "reload");
  assert(ok, "reload mode should not require a pre-step");
});

Deno.test("Refresh - none mode has no pre-step", { timeout: 5000 }, async () => {
  const ok = await refreshPreStep("/tmp/test.lyx", "none");
  assert(ok, "none mode should not require a pre-step");
});

// DL82 (test_report_34 Finding 9): LyX's server flushes ALL clients' responses
// into every .out instance; lq must correlate responses by client name, not
// position, and ignore stale lines from other clients.
Deno.test("Refresh - .out responses filtered by client name", () => {
  const mine = "lq1785486000000";
  const other = "lq1785485999000";

  // Stale lines from other clients must be ignored even if they appear after
  // this client's line in the buffer.
  const data = [
    `ERROR:${other}:buffer-reload:Command disabled`,
    `INFO:${mine}:buffer-write:`,
    `ERROR:${other}:buffer-switch X:Document not loaded`,
  ].join("\n");
  assertEquals(
    filterResponses(data, mine),
    `INFO:${mine}:buffer-write:`,
    "must return the last line for THIS client",
  );

  // An ERROR line for this client is also returned (caller decides success).
  assertEquals(
    filterResponses(`ERROR:${mine}:buffer-reload:Command disabled`, mine),
    `ERROR:${mine}:buffer-reload:Command disabled`,
  );

  // No line for this client yet → null (keep polling).
  assertEquals(filterResponses(data, "lq1785487000000"), null);
});

// DL82 (test_report_34 Finding 9): the matching line must be matched by client
// name — a response for a client whose name is a PREFIX of another must not be
// conflated.
Deno.test("Refresh - client-name filter is exact (not prefix)", () => {
  const data = `INFO:lq123:buffer-write:\nINFO:lq1234:buffer-write:`;
  assertEquals(filterResponses(data, "lq123"), `INFO:lq123:buffer-write:`);
  assertEquals(filterResponses(data, "lq1234"), `INFO:lq1234:buffer-write:`);
});

// Dev log 128: the Windows pipe transport must send the colon-separated
// func:arg form so a drive letter in an absolute path doesn't truncate the
// func field (Server.cpp Server::callback parses the func field up to the next
// ':' and the arg field up to '\n'). Overturns dev log 28's "no escape
// mechanism" conclusion.
Deno.test("Refresh - pipe command uses colon-separated func:arg form", () => {
  const c = "lq123";
  // No argument → bare func, no trailing colon (unchanged).
  assertEquals(buildPipeCommand(c, "buffer-write"), `LYXCMD:${c}:buffer-write\n`);
  // Windows absolute path → colon form keeps the drive letter in the argument.
  assertEquals(
    buildPipeCommand(c, "buffer-switch C:\\Users\\Shifu\\file.lyx"),
    `LYXCMD:${c}:buffer-switch:C:\\Users\\Shifu\\file.lyx\n`,
  );
  // Path containing spaces is preserved (argument runs to '\n').
  assertEquals(
    buildPipeCommand(c, "buffer-switch C:\\Users\\Shifu\\LyX 2.5\\file.lyx"),
    `LYXCMD:${c}:buffer-switch:C:\\Users\\Shifu\\LyX 2.5\\file.lyx\n`,
  );
});

Deno.test("Refresh - mock server: save-reload returns ok (Unix socket)", { ignore: Deno.build.os === "windows", timeout: 15000 }, async () => {
  // Backlog 17 / dev log 87 F1: a controllable server behind the $LYXSOCKET
  // override proves refreshPreStep(file, "save-reload") returns "ok" — not just
  // "disconnect" (no LyX) — and that the expected LFUNs were received.
  // Deterministic and GUI-free.
  //
  // Guarded to Unix: Deno.listen with the unix transport is unsupported on
  // Windows (op_net_listen_unix), so the Windows named-pipe path cannot be
  // mocked in-test — it is covered by live tests instead (dev log 87 F1/F10).
  const tmp = Deno.makeTempDirSync();
  const socketPath = `${tmp}/lyxsocket`;
  const filePath = `${tmp}/doc.lyx`;
  const originalLyxSocket = Deno.env.get("LYXSOCKET");
  await Deno.writeTextFile(filePath,
    "#LyX 2.5 created this file.\n" +
    "\\begin_document\n\\begin_header\n\\end_header\n" +
    "\\begin_body\n\\begin_layout Standard\nhi\n\\end_layout\n\\end_body\\end_document\n");
  try {
    Deno.env.set("LYXSOCKET", socketPath);
    const listener = Deno.listen({ path: socketPath, transport: "unix" });
    const decoder = new TextDecoder();
    const encoder = new TextEncoder();
    const buf = new Uint8Array(4096);
    let sawBufferSwitch = false;
    let sawBufferWrite = false;
    const serverDone = (async () => {
      const conn = await listener.accept();
      let data = "";
      try {
        while (true) {
          const n = await conn.read(buf);
          if (n === null) break;
          data += decoder.decode(buf.subarray(0, n));
          let nl: number;
          while ((nl = data.indexOf("\n")) !== -1) {
            const line = data.slice(0, nl).trim();
            data = data.slice(nl + 1);
            if (line.startsWith("HELLO:")) {
              await conn.write(encoder.encode("HELLO:\n"));
            } else if (line.startsWith("LYXCMD:")) {
              const lfun = line.slice("LYXCMD:".length);
              if (lfun.includes("buffer-switch")) sawBufferSwitch = true;
              if (lfun.includes("buffer-write")) sawBufferWrite = true;
              await conn.write(encoder.encode(`INFO:${lfun}:\n`));
            } else if (line.startsWith("BYE:")) {
              break;
            }
          }
        }
      } finally {
        try { conn.close(); } catch { /* ignore */ }
        listener.close();
      }
    })();
    const status = await refreshPreStep(filePath, "save-reload");
    await serverDone;
    assertEquals(status, "ok", "save-reload must report ok against a responding server");
    assertEquals(sawBufferSwitch, true, "server must receive buffer-switch before the save");
    assertEquals(sawBufferWrite, true, "server must receive buffer-write");
  } finally {
    if (originalLyxSocket === undefined) Deno.env.delete("LYXSOCKET");
    else Deno.env.set("LYXSOCKET", originalLyxSocket);
    try { Deno.removeSync(tmp, { recursive: true }); } catch { /* ignore */ }
  }
});

File: tests/refresh_test.rs
//! LyXServer refresh helpers (Deno `tests/refresh_test.ts`).

mod common;

use common::{IsolatedHome, WorkDir, json_stdout, run_cli_with_env};
use lq::{RefreshMode, RefreshPreStep, build_pipe_command, filter_responses, refresh_pre_step};
use std::sync::Mutex;

static ENV_LOCK: Mutex<()> = Mutex::new(());

fn lock_lyxsocket() -> std::sync::MutexGuard<'static, ()> {
    ENV_LOCK.lock().unwrap_or_else(|e| e.into_inner())
}

#[test]
fn refresh_save_reload_pre_step_returns_a_status() {
    let _guard = lock_lyxsocket();
    let status = refresh_pre_step("/tmp/test.lyx", RefreshMode::SaveReload);
    assert!(
        matches!(
            status,
            RefreshPreStep::Ok
                | RefreshPreStep::Disconnect
                | RefreshPreStep::Unconfirmed
                | RefreshPreStep::Error
        ),
        "refresh_pre_step must return a status, got: {status:?}"
    );
}

#[test]
fn refresh_reload_mode_has_no_pre_step() {
    assert_eq!(
        refresh_pre_step("/tmp/test.lyx", RefreshMode::Reload),
        RefreshPreStep::Ok
    );
}

#[test]
fn refresh_none_mode_has_no_pre_step() {
    assert_eq!(
        refresh_pre_step("/tmp/test.lyx", RefreshMode::None),
        RefreshPreStep::Ok
    );
}

#[test]
fn refresh_out_responses_filtered_by_client_name() {
    let mine = "lq1785486000000";
    let other = "lq1785485999000";
    let data = [
        format!("ERROR:{other}:buffer-reload:Command disabled"),
        format!("INFO:{mine}:buffer-write:"),
        format!("ERROR:{other}:buffer-switch X:Document not loaded"),
    ]
    .join("\n");
    assert_eq!(
        filter_responses(&data, mine).as_deref(),
        Some("INFO:lq1785486000000:buffer-write:")
    );
    assert_eq!(
        filter_responses("ERROR:lq1785486000000:buffer-reload:Command disabled", mine).as_deref(),
        Some("ERROR:lq1785486000000:buffer-reload:Command disabled")
    );
    assert_eq!(filter_responses(&data, "lq1785487000000"), None);
}

#[test]
fn refresh_client_name_filter_is_exact_not_prefix() {
    let data = "INFO:lq123:buffer-write:\nINFO:lq1234:buffer-write:";
    assert_eq!(
        filter_responses(data, "lq123").as_deref(),
        Some("INFO:lq123:buffer-write:")
    );
    assert_eq!(
        filter_responses(data, "lq1234").as_deref(),
        Some("INFO:lq1234:buffer-write:")
    );
}

#[test]
fn refresh_pipe_command_uses_colon_separated_func_arg_form() {
    let c = "lq123";
    assert_eq!(
        build_pipe_command(c, "buffer-write"),
        "LYXCMD:lq123:buffer-write\n"
    );
    assert_eq!(
        build_pipe_command(c, "buffer-switch C:\\Users\\Shifu\\file.lyx"),
        "LYXCMD:lq123:buffer-switch:C:\\Users\\Shifu\\file.lyx\n"
    );
    assert_eq!(
        build_pipe_command(c, "buffer-switch C:\\Users\\Shifu\\LyX 2.5\\file.lyx"),
        "LYXCMD:lq123:buffer-switch:C:\\Users\\Shifu\\LyX 2.5\\file.lyx\n"
    );
}

#[test]
fn cli_init_refresh_warns_when_lyxsocket_unreachable() {
    let home = IsolatedHome::new();
    let layouts = WorkDir::new();
    let work = WorkDir::new();
    let bogus = work
        .path()
        .join(format!("no-lyx-socket-{}", std::process::id()));
    let out = run_cli_with_env(
        &[
            "init",
            "--global",
            "--layouts-dir",
            layouts.path().to_str().expect("layout path is utf-8"),
            "--refresh",
            "save-reload",
        ],
        &home,
        work.path(),
        &[("LYXSOCKET", bogus.to_str().expect("socket path is utf-8"))],
    );
    assert_eq!(out.code, 0, "{}", out.stdout);
    let v = json_stdout(&out);
    assert_eq!(v["data"]["refresh"], "save-reload");
    let warnings = v["warnings"]
        .as_array()
        .expect("init JSON includes warnings");
    assert!(
        warnings.iter().any(|w| w.as_str().is_some_and(|s| {
            s.contains("requires a running LyX instance with LyXServer enabled")
                && s.contains("server could not be reached")
        })),
        "expected unreachable-server warning, got {warnings:?}"
    );
}

#[cfg(unix)]
mod unix_mock {
    use super::*;
    use std::io::{Read, Write};
    use std::os::unix::net::UnixListener;

    #[test]
    fn refresh_mock_server_save_reload_returns_ok_unix_socket() {
        let _guard = super::lock_lyxsocket();
        let tmp = std::env::temp_dir().join(format!(
            "lq_lyxsocket_{}_{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_nanos())
                .unwrap_or(0)
        ));
        std::fs::create_dir_all(&tmp).expect("mock temp dir");
        let socket_path = tmp.join("lyxsocket");
        let file_path = tmp.join("doc.lyx");
        std::fs::write(
            &file_path,
            "#LyX 2.5 created this file.\n\
\\begin_document\n\\begin_header\n\\end_header\n\
\\begin_body\n\\begin_layout Standard\nhi\n\\end_layout\n\\end_body\\end_document\n",
        )
        .expect("write mock lyx");

        let listener = UnixListener::bind(&socket_path).expect("bind mock lyxsocket");
        let server = std::thread::spawn(move || {
            let (mut conn, _) = listener.accept().expect("accept mock client");
            let mut data = Vec::new();
            let mut buf = [0u8; 4096];
            let mut saw_buffer_switch = false;
            let mut saw_buffer_write = false;
            loop {
                let n = match conn.read(&mut buf) {
                    Ok(0) | Err(_) => break,
                    Ok(n) => n,
                };
                data.extend_from_slice(&buf[..n]);
                while let Some(nl) = data.iter().position(|&b| b == b'\n') {
                    let line = String::from_utf8_lossy(&data[..nl]).trim().to_string();
                    data.drain(..=nl);
                    if line.starts_with("HELLO:") {
                        let _ = conn.write_all(b"HELLO:\n");
                    } else if let Some(lfun) = line.strip_prefix("LYXCMD:") {
                        if lfun.contains("buffer-switch") {
                            saw_buffer_switch = true;
                        }
                        if lfun.contains("buffer-write") {
                            saw_buffer_write = true;
                        }
                        let reply = format!("INFO:{lfun}:\n");
                        let _ = conn.write_all(reply.as_bytes());
                    } else if line.starts_with("BYE:") {
                        return (saw_buffer_switch, saw_buffer_write);
                    }
                }
            }
            (saw_buffer_switch, saw_buffer_write)
        });

        let previous = std::env::var("LYXSOCKET").ok();
        // SAFETY: ENV_LOCK is held; this test target is the only mutator of LYXSOCKET.
        unsafe {
            std::env::set_var("LYXSOCKET", &socket_path);
        }
        let status = refresh_pre_step(
            file_path.to_str().expect("file path is utf-8"),
            RefreshMode::SaveReload,
        );
        match previous {
            Some(value) => unsafe { std::env::set_var("LYXSOCKET", value) },
            None => unsafe { std::env::remove_var("LYXSOCKET") },
        }

        let (saw_switch, saw_write) = server.join().expect("mock server thread");
        let _ = std::fs::remove_dir_all(&tmp);
        assert_eq!(status, RefreshPreStep::Ok);
        assert!(
            saw_switch,
            "server must receive buffer-switch before the save"
        );
        assert!(saw_write, "server must receive buffer-write");
    }
}
